### PR TITLE
[main][feat] add contract single asset  partial close withdraw minimize 

### PR DESCRIPTION
**Pull request by sundayalpaca**
…ning

## Description
feat: add contract single asset  partial close withdraw minimize trai…

### CODE REVIEW

File: contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+      uint256 farmingTokenToLiquidate,
+      uint256 toRepaidBaseTokenDebt
+    ) = abi.decode(data, (uint256, uint256, uint256));
+    IWorker02 worker = IWorker02(msg.sender);
+    address baseToken = worker.baseToken();
+    address farmingToken = worker.farmingToken();
+    // 2. Approve router to do their stuffs
+    farmingToken.safeApprove(address(router), uint256(-1));
+    // 3. Swap from farming token -> base token according to worker's path
+    uint256 farmingTokenBalance = farmingToken.myBalance();
```
Comment (slowcheetah69) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol:80]: The farmingToken balance should be checked here if it is enough to pay back the debt.
```suggestion
require(
      farmingToken.myBalance() >= farmingTokenToLiquidate,
      "PancakeswapV2RestrictedSingleAssetStrategyPartialCloseLiquidate::execute:: insufficient farmingToken received from worker"
    );
```
see https://github.com/alpaca-finance/bsc-alpaca-contract/pull/70/files

File: contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    uint256 farmingTokenToBeRepaidDebt = 0;
+    if (toRepaidBaseTokenDebt > 0) {
+        uint256[] memory farmingTokenToBeRepaidDebts = router.getAmountsIn(toRepaidBaseTokenDebt, worker.getReversedPath());
+        farmingTokenToBeRepaidDebt = farmingTokenToBeRepaidDebts[0];
+        require(farmingTokenToLiquidate > farmingTokenToBeRepaidDebts[0], "PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading::execute:: not enough to pay back debt");
```
Comment (ricardo-alpaca) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol:85]: What if ``farmingTokenToLiquidate == farmingTokenToBeRepaidDebts[0]``?
Should we allow it to pass?

File: contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import "hardhat/console.sol";
+
+
+contract PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading is OwnableUpgradeSafe, ReentrancyGuardUpgradeSafe, IStrategy {
+  using SafeToken for address;
+  using SafeMath for uint256;
+
+  IPancakeFactory public factory;
+  IPancakeRouter02 public router;
+  address public wNative;
```
Comment (ricardo-alpaca) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol:39]: In ``PancakeswapV2RestrictedSingleAssetStrategyWithdrawMinimizeTrading.sol``, this variable was declared as
``IWETH public wbnb;`` which seems better to me as it is defined as IWETH type. However, ``PancakeswapV2RestrictedSingleAssetStrategyWithdrawMinimizeTrading.sol`` also used ``address public wNative;``. Perhaps we could be more consistent on this. Try discussing with @abraham-alpaca on this.

**Review by ricardo-alpaca — CHANGES_REQUESTED** [review]
There are some minor inconsistencies and one questionable check. Please discuss with the other alpaca to resolve them.

File: contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  using SafeToken for address;
+  using SafeMath for uint256;
+
+  IPancakeFactory public factory;
+  IPancakeRouter02 public router;
+  address public wbnb;
+  mapping(address => bool) public okWorkers;
+  IWNativeRelayer public wNativeRelayer;
+
+  event PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading(
```
Comment (ricardo-alpaca) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol:43]: Change event name to `PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTradingEvent`

**Review by ricardo-alpaca — APPROVED** [review]

File: contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+import "@pancakeswap-libs/pancake-swap-core/contracts/interfaces/IPancakePair.sol";
+
+
+import "../../apis/pancake/IPancakeRouter02.sol";
+import "../../interfaces/IStrategy.sol";
+import "../../../utils/SafeToken.sol";
+import "../../../utils/AlpacaMath.sol";
+import "../../interfaces/IWorker02.sol";
+import "../../interfaces/IWNativeRelayer.sol";
+import "hardhat/console.sol";
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol:30]: remove hardhat/console.sol please

File: contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+  event PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTradingEvent(
+    address indexed baseToken,
+    address indexed farmToken,
+    uint256 amounToLiquidate,
+    uint256 amountToRepayDebt
+  );
+
+  // @notice require that only allowed workers are able to do the rest of the method call
+  modifier onlyWhitelistedWorkers() {
+    require(okWorkers[msg.sender], "PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading::onlyWhitelistedWorkers:: bad worker");
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol:52]: setup solhint/lint

File: contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+  /// @param data Extra calldata information passed along to this strategy.
+  function execute(address user, uint256 /*debt*/, bytes calldata data)
+    external
+    override
+    onlyWhitelistedWorkers
+    nonReentrant
+  {
+    // 1. minFarmingTokenAmount for validating a farmingToken amount after leaving the stake.
+    (
+      uint256 minFarmingTokenAmount,
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol:77]: reorder the variables.
- farmingTokenToLiquidate
- toRepaidBaseTokenDebt
- minFarmingTokenAmount

File: contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+  /// @dev Execute worker strategy. take farmingToken, return farmingToken + basetoken that is enough to repay the debt
+  /// @param data Extra calldata information passed along to this strategy.
+  function execute(address user, uint256 /*debt*/, bytes calldata data)
+    external
+    override
+    onlyWhitelistedWorkers
+    nonReentrant
+  {
+    // 1. minFarmingTokenAmount for validating a farmingToken amount after leaving the stake.
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol:75]: Update comment here

File: contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    IWorker02 worker = IWorker02(msg.sender);
+    address baseToken = worker.baseToken();
+    address farmingToken = worker.farmingToken();
+    // 2. Approve router to do their stuffs
+    farmingToken.safeApprove(address(router), uint256(-1));
+    // 3. Swap from farming token -> base token according to worker's path
+    uint256 farmingTokenBalance = farmingToken.myBalance();
+    require(farmingTokenBalance >= farmingTokenToLiquidate, "PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading::execute:: insufficient farmingToken received from worker");
+    uint256 farmingTokenToBeRepaidDebt = 0;
+    if (toRepaidBaseTokenDebt > 0) {
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol:90]: add comment above this step for readability

**Review by spicysquid168 — CHANGES_REQUESTED** [review]

File: contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        SafeToken.safeTransfer(farmingToken, address(wNativeRelayer), farmingTokenBalanceToBeSentToTheUser);
+        wNativeRelayer.withdraw(farmingTokenBalanceToBeSentToTheUser);
+        SafeToken.safeTransferETH(user, farmingTokenBalanceToBeSentToTheUser);
+      } else {
+        SafeToken.safeTransfer(farmingToken, user, farmingTokenBalanceToBeSentToTheUser);
+      }
+    }
+    // 6. Reset approval for safety reason
+    farmingToken.safeApprove(address(router), 0);
+    uint256 lpTokenToLiquidate = 0;
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol:112]: do we need it here? I think we don't need this variable here.

File: contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    emit PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTradingEvent(
+      baseToken,
+      farmingToken,
+      lpTokenToLiquidate,
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol:116]: Should be `farmingTokenToLiquidate`

File: contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+    if (farmingTokenBalanceToBeSentToTheUser > 0) {
+      if (farmingToken == address(wbnb)) {
+        SafeToken.safeTransfer(farmingToken, address(wNativeRelayer), farmingTokenBalanceToBeSentToTheUser);
+        wNativeRelayer.withdraw(farmingTokenBalanceToBeSentToTheUser);
+        SafeToken.safeTransferETH(user, farmingTokenBalanceToBeSentToTheUser);
+      } else {
+        SafeToken.safeTransfer(farmingToken, user, farmingTokenBalanceToBeSentToTheUser);
+      }
+    }
+    // 6. Reset approval for safety reason
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol:110]: need to transfer `farmingToken.safeTransfer(farmingTokenBalance.sub(farmingTokenToRepaidDebt))` back to worker to return the leftover amount from partial close

**Review by spicysquid168 — CHANGES_REQUESTED** [review]

File: contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+    nonReentrant
+  {
+    // 1.1 farmingTokenToLiquidate for farmingToken amount that user want to liquidate.
+    // 1.2 toRepaidBaseTokenDebt for baseToken amount that user want to repaid debt.
+    // 1.3 minFarmingTokenAmount for validating a farmingToken amount.
+    (
+      uint256 farmingTokenToLiquidate,
+      uint256 toRepaidBaseTokenDebt,
+      uint256 minFarmingTokenAmount
+    ) = abi.decode(data, (uint256, uint256, uint256));
```
Comment (spicysquid168) [inline contracts/6/protocol/strategies/pancakeswapV2-restricted-single-asset/PancakeswapV2RestrictedSingleAssetStrategyPartialCloseWithdrawMinimizeTrading.sol:80]: add require(toRepaidBaseTokenDebt <= debt)

**Review by spicysquid168 — CHANGES_REQUESTED** [review]

**Review by spicysquid168 — DISMISSED** [review]

**Review by almost-llamas — DISMISSED** [review]

File: test/CakeMaxiWorker.test.ts [hunk 14]
```diff
@@ -854,6 +864,63 @@ describe('CakeMaxiWorker', () => {
...
+                ['uint256'],
+                [ethers.utils.parseEther('0')]
+              )
+            ],
+          )
+        )
+        let userInfo = await masterChef.userInfo(0, cakeMaxiWorkerNative.address)
+        expect(userInfo[0]).to.eq(ethers.utils.parseEther('0.00907024323709934'))
+        expect(await cakeMaxiWorkerNative.shares(0)).to.eq(ethers.utils.parseEther('0.00907024323709934'))
+        // Alice call minimize trading strategy to close her position
```
Comment (spicysquid168) [inline test/CakeMaxiWorker.test.ts:892]: Add more context here. How much Alice wish to close position? How much she repay the debt? etc.

File: test/CakeMaxiWorker.test.ts [hunk 15]
```diff
@@ -854,6 +864,63 @@ describe('CakeMaxiWorker', () => {
...
+              ethers.utils.defaultAbiCoder.encode(
+                ['uint256','uint256','uint256'],
+                [ethers.utils.parseEther('0.005'),ethers.utils.parseEther('0.02'),ethers.utils.parseEther('0')]
+              )
+            ],
+          )
+        )
+        // 0.1 - 0.00907024323709934 FTOKEN = 1.1 WBNB
+        // 0.09092975676290066 FTOKEN =  1.1 WBNB
+        // x FTOKEN = (x * 0.9975 * 1.1) / (0.09092975676290066 + (x * 0.9975)) = 0.05 WBNB
```
Comment (spicysquid168) [inline test/CakeMaxiWorker.test.ts:907]: where does 0.05 WBNB comes from? Can you elaborate the math more in the comment? i.e. Finding how much 0.05WBNB worth in FTOKEN: equation

File: test/CakeMaxiWorker.test.ts [hunk 16]
```diff
@@ -854,6 +864,63 @@ describe('CakeMaxiWorker', () => {
...
+        // x = 0.004340840518577427
+        // x FTOKEN = (x * 0.9975 * 1.1) / (0.09092975676290066 + (x * 0.9975)) = 0.02 WBNB
+        // x = 0.0016881046461134437
```
Comment (spicysquid168) [inline test/CakeMaxiWorker.test.ts:910]: same here. Elaborate more. Something like Finding how many FTOKEN needed to enough to pay 0.02 debt: equation

**Review by spicysquid168 — CHANGES_REQUESTED** [review]

File: test/CakeMaxiWorker.test.ts [hunk 17]
```diff
@@ -854,6 +864,73 @@ describe('CakeMaxiWorker', () => {
...
+            ['address', 'bytes'],
+            [stratPartialCloseMinimize.address, 
+              ethers.utils.defaultAbiCoder.encode(
+                ['uint256','uint256','uint256'],
+                [ethers.utils.parseEther('0.005'),ethers.utils.parseEther('0.02'),ethers.utils.parseEther('0')]
+              )
+            ],
+          )
+        )
+        // alice want to repaid bedt only 0.02 BToken
```
Comment (spicysquid168) [inline test/CakeMaxiWorker.test.ts:917]: debt

File: test/CakeMaxiWorker.test.ts [hunk 18]
```diff
@@ -854,6 +864,73 @@ describe('CakeMaxiWorker', () => {
...
+              )
+            ],
+          )
+        )
+        let userInfo = await masterChef.userInfo(0, cakeMaxiWorkerNative.address)
+        expect(userInfo[0]).to.eq(ethers.utils.parseEther('0.00907024323709934'))
+        expect(await cakeMaxiWorkerNative.shares(0)).to.eq(ethers.utils.parseEther('0.00907024323709934'))
+
+        // Alice call partial close minimize trading strategy to close some her position
+        // if Alice bedt's is 50% of her position which is 0.1 wbnb
```
Comment (spicysquid168) [inline test/CakeMaxiWorker.test.ts:895]: debt

Comment (spicysquid168) [reply test/CakeMaxiWorker.test.ts:895]: Alice has 0.05 WBNB as debt. Finding how much 0.05 WBNB worth in FTOKEN:
Equation..

**Review by spicysquid168 — CHANGES_REQUESTED** [review]

**Review by spicysquid168 — APPROVED** [review]

**Review by almost-llamas — APPROVED** [review]

**Pull request merged**